Fixes VaultTransitSecretEngine.readKey to work with symmetric & asymmetric keys

kdubb opened the pull request:
This fixes the current failure to read asymmetric keys while adding more information in a backwards compatible and futue proof format.

Fixes #16988

### DTO Changes
* Changes `VaultTransitReadKeyData.keys` to be a list of `VaultTransitKeyVersionData`(which can contain any key data)
* Uses custom deserializer on `VaultTransitReadKeyData.keys` to allow `VaultTransitKeyVersionData` values to be deserialized from an integer timestamp or a map of info values.
* Added missing `latest_version`, `min_available_version` and `type` fields to `VaultTransitReadKeyData`

### API Changes
* Made `VaultTransitKeyDetail` abstract, subclasses for symmetric (`VaultTransitSymmetricKeyDetail`) and asymmetric (`VaultTransitAsymmetricKeyDetail`) keys are provided.
* Added abstract `versions` property to `VaultTransitKeyDetail` which is a `Map<String, ? extends VaultTransitKeyVersion>`, subclasses of `VaultTransitKeyDetail` provide a concrete `versions` property with an unambiguous type.
* * E.g. `VaultTransitAsymmetricKeyDetail.versions` is defined as `Map<String, VaultTransitAsymmetricKeyVersion>`
* * Allows users to test the key detail type once (via `instanceof`) and downcast to a detailt type that has a specifically typed list of versions.
* `VaultTransitKeyVersion` is abstract also, with subclasses for asymmetric (`VaultTransitAsymmetricKeyVersion`) and symmetric (`VaultTransitSymmetricKeyVersion`) keys
* Added `latestVersion`, `minAvailableVersion` and `type` properties to `VaultTransitKeyDetail`
* Changes are additive and therefore backwards compatible
* Deprecates the `keys` property as `versions` provides the same information in a more easily accessible format and is future proof for adding more properties as Vault makes changes to the API.

### Tests
* Added integration test cases for reading ECDSA, RSA and AES keys

### Miscellaneous
* Added `JavaTimeModule` to the mapper used for vault responses.

kdubb committed:
Fixes VaultTransitSecretEngine.readKey to work with symmetric & asymmetric keys

This fixes the current failure to read asymmetric keys while adding more information in a backwards compatible and futue proof format.

Fixes #16988

* Changes `VaultTransitReadKeyData.keys` to be a list of `VaultTransitKeyVersionData`(which can contain any key data)
* Uses custom deserializer on `VaultTransitReadKeyData.keys` to allow `VaultTransitKeyVersionData` values to be deserialized from an integer timestamp or a map of info values.
* Added missing `latest_version`, `min_available_version` and `type` fields to `VaultTransitReadKeyData`

* Made `VaultTransitKeyDetail` abstract, subclasses for symmetric (`VaultTransitSymmetricKeyDetail`) and asymmetric (`VaultTransitAsymmetricKeyDetail`) keys are provided.
* Added abstract `versions` property to `VaultTransitKeyDetail` which is a `Map<String, ? extends VaultTransitKeyVersion>`, subclasses of `VaultTransitKeyDetail` provide a concrete `versions` property with an unambiguous type.
* * E.g. `VaultTransitAsymmetricKeyDetail.versions` is defined as `Map<String, VaultTransitAsymmetricKeyVersion>`
* * Allows users to test the key detail type once (via `instanceof`) and downcast to a detailt type that has a specifically typed list of versions.
* `VaultTransitKeyVersion` is abstract also, with subclasses for asymmetric (`VaultTransitAsymmetricKeyVersion`) and symmetric (`VaultTransitSymmetricKeyVersion`) keys
* Added `latestVersion`, `minAvailableVersion` and `type` properties to `VaultTransitKeyDetail`
* Changes are additive and therefore backwards compatible
* Deprecates the `keys` property as `versions` provides the same information in a more easily accessible format and is future proof for adding more properties as Vault makes changes to the API.

* Added integration test cases for reading ECDSA, RSA and AES keys

* Added `JavaTimeModule` to the mapper used for vault responses.

vsevel commented:
great! I will have a look tomorrow.

kdubb commented:
@vsevel Vault Transit docs still say it's in _preview_.  Given that we can nix backward compatibility, It might good to just remove the `keys` property altogether. The data is redundant and the epoch integer is less useful that the `createTime`  property on version values.

sberyozkin commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/VaultTransitAsymmetricKeyVersion.java:
```diff
@@ -0,0 +1,25 @@
+package io.quarkus.vault;
+
+public class VaultTransitAsymmetricKeyVersion extends VaultTransitKeyVersion {
```
`Version` reads redundant, the key representations are read, so why not just have `VaultTransitAsymmetricKey`

kdubb replied:
Vault Transit keys can be rotated creating many versions of each key.  The `keys` property is actually a map with the map key being the version number (as a string) and the value describing each version of the key or the integer timestamp that version was created. This is a nomenclature issues created by the Vault API, which is why my new property is named `versions`.

sberyozkin replied:
@kdubb - version can be a property of a given key representation ?

kdubb replied:
Yes.  A  `VaultTransitKeyVersion` _is always_ exposed via the `versions` map property of a `VaultTransitKeyDetail`.

sberyozkin replied:
In that case, why to keep `Version` in the name of the class designed to provide a key representation ?

kdubb replied:
The "class designed to provided a key representation" is `VaultTransitKeyDetail` and existed before this PR. A Vault "key representation" can rotated which creates multiple versions. `VaultTransitKeyVersion` provides data about a _single_ version of the key. So it is named for exactly the data it is providing.

sberyozkin replied:
@kdubb

This PT turns `VaultTransitKeyDetail` into a base container of key representations where each representation has `Version` in its name, I'm finding it confusing

kdubb replied:
Prior to this PR, `VaultTransitKeyDetail` was already a base container of key version representations, that what the `keys` property is. Previosuly the version representation was only a single timestamp of when the version was created.

Instead of altering the `keys` property (for backward compatibility), I've added the `versions` property that simply provides more data about each version. `VaultTransitKeyVersion` is the container for all data about a specific version of a key.

If we were to drop the "Version" from `VaultTransitKeyVersion` the name would become `VaultTransitKey`; which would be very much misnamed, as it does **_not_** represent a Vault key. It only represents data about a specific version of a Vault key.

kdubb replied:
Also, the name of the `keys` property comes from the Vault API and I would argue that it is ambiguously named because it is a map of version numbers to version details; `versions` is a better name and we should use it.

sberyozkin replied:
OK, you obviously understand the details of the Vault API better - my intention is not to interfere but only to try to make sure the potential users would not find it confusing - but I'll let Vincent and yourself to finalize - thank you

vsevel replied:
introducing a new `versions` field makes sense, as the vault actually says: _The keys attribute lists each version of the key,_
I am ok also with the `VaultTransitKeyVersion` terminology.

sberyozkin commented on integration-tests/vault/src/test/java/io/quarkus/vault/VaultTransitITCase.java:
```diff
@@ -314,4 +317,114 @@ public void adminKey() {
+        assertTrue(mykey.getVersions().containsKey("1"));
+        assertNotNull(mykey.getVersions().get("1").getCreationTime());
+        assertTrue(mykey.getVersions().get("1") instanceof VaultTransitAsymmetricKeyVersion);
+        assertNotNull(((VaultTransitAsymmetricKeyVersion) mykey.getVersions().get("1")).getPublicKey());
```
This does not look good IMHO

sberyozkin replied:
There could be only symmetric and asymmetric keys - so a more user friendly approach is to provide an option to retrieve symmetric keys and asymmetric keys as opposed to expecting the users to cast

kdubb replied:
This was for testing. The new detail and version objects make it easy to test and access versions without excessive casting.

For example...
```java
if (mykey instanceof VaultTransitAsymmetricKeyDetail) {
  VaultTransitAsymmetricKeyDetail myAsymKey = (VaultTransitAsymmetricKeyDetail)mykey;
  // myAsymKey.versions is typed to asymmetric versions
  for (VaultTransitAsymmetricKeyVersion version : myAsymKey.versions) {
    ... 
  }
} else {
  VaultTransitSymmetricKeyDetail mySymKey = (VaultTransitSymmetricKeyDetail)mykey;
  // mySymKey.versions is typed to symmetric versions
  for (VaultTransitSymmetricKeyVersion version : mySymKey.versions) {
    ... 
  }
}
```

With Java 14... it's even easier
```java
if (mykey instanceof VaultTransitAsymmetricKeyDetail myAsymKey) {
  for (VaultTransitAsymmetricKeyVersion version : myAsymKey.versions) {
    ... 
  }
}
```

kdubb replied:
To be clear all versions in a detail object will be of the same type. The Vault API will not allow mixing asymmetric and symmetric key versions. So, with your suggestion, you'd have two separate maps of versions and one would always be empty. You'd have to check the size of the maps to see which one is not empty and use that one. That's more work and more confusing than checking the detail object's type and casting the detail object to get the correctly typed versions.

sberyozkin replied:
I'm sorry but it does not look easy to me. This is exactly why I doubted this approach in the issue discussion `instranceof plus cast` - it is not really how I'd expect Quarkus Vault users accessing the keys in the new version of this API

kdubb replied:
🤷‍♂️ That's a problem you seem to have with Java.

Again, your suggestion appears to be to create _**two**_ maps of versions, one for symmetric versions and one or asymmetric versions. That is **not** how the data is organized because each `VaultTransitKeyDetail` can only contain one type of versions. You'd have two maps and one would always be completely empty. In this scenario, without using`instanceof` how would a user check if it's an asymmetric or symmetric key?

sberyozkin replied:
Perhaps a single map can be used but in addition to the users being able to access this map they can have options for getting specific representations ? How about adding a few more helpers to get a list of asymmetric/symmetric keys ? That would be a reasonable compromise IMHO

sberyozkin replied:
Also, I was thinking yesterday, should the users have access to `PublicKey` instances as opposed to them having to parse PEM content ? May be both options should be available...

kdubb replied:
> Perhaps a single map can be used but in addition to the users being able to access this map they can have options for getting specific representations ?

The user must still first determine if the key is symmetric or asymmetric prior to asking for specific representations. If the user does that via `instanceof`, then they have access to the specific (and type correct) representations via the detail subclasses.

It might help to reexamine the classes `VaultTransitKeyDetail` and it's subclasses `VaultTransitAsymmetricKeyDetail` and `VaultTransitSymmetricKeyDetail`.  Once a user has determined wether a key detail object is symmetric or asymmetric (I'm using `instanceof` to determine this), a single cast gives users exactly what you are asking for, specific representations of the versions.

kdubb replied:
> Also, I was thinking yesterday, should the users have access to `PublicKey` instances as opposed to them having to parse PEM content ? May be both options should be available...

I like this idea but there's a bit of a problem.  There's no good PEM parser in Java security so to support this we'd have to bring in BouncyCastle or roll our own PEM parser.

Also, `VaultTransitKeyExportDetail` represents its PEM private key versions as `String`s. So, currently there is consistency.

sberyozkin replied:
I don't agree with the `instanceof/cast` approach - IMHO it is not user friendly at all - but I'll let Vincent decide.

I don't think a 3rd party parser is needed but agreed representing keys are strings is fine for now

vsevel replied:
catching up on the discussions.

I am not sure the `VaultTransitKeyDetail` is bringing a lot of value with the 2 subclasses. or at the very least the abstract superclass should introduce field `Map<String, T> versions`, with `VaultTransitKeyDetail` being parameterized with `T extends VaultTransitKeyDetail`. then subclasses should defined `T` consistently: `VaultTransitSymmetricKeyVersion` for `VaultTransitSymmetricKeyDetail`, and `VaultTransitAsymmetricKeyVersion` for `VaultTransitAsymmetricKeyDetail`.

the other option is to keep a single class `VaultTransitKeyDetail` like before , and add field `Map<String, ? extends VaultTransitKeyVersion> versions`. I have a slight preference for that solution, since after you do work described just above, you will end up with empty `VaultTransitSymmetricKeyDetail` and `VaultTransitAsymmetricKeyDetail` classes, serving essentially as markers to get the appropriate parametrized types for the `versions` map.

what do you all think?

kdubb replied:
First, we need "some" method for users to check if it's a symmetric or asymmetric key. We should _not_ require them to use the `type` property because as Vault adds more key types, the `type` field domain will grow and become a maintenance headache. So, we need a simple check for asymmetric versus symmetric.

We can do that with subclasses (as I have done) or we can do it with another explicit field (e.g. `kind` or `isAsymmetric`). The method I have implemented using subclasses has advantages in that both Java 14+ and Kotlin have pattern matching for this exact scenario.  In Java 14+ `instanceof` checks can easily elide the cast and Kotlin automatically casts the checked values. With an explicit field (e.g. `kind`) the user would be required to check the field and then perform an explicit cast anyway in both languages.

##### Examples for `instanceof` method in Java & Kotlin

`instanceof` (Java 14+)
```java
if (mykey instanceof VaultTransitAsymmetricKeyDetail myAsymKey) {
  for (var version : myAsymKey.versions) {
    ... 
  }
}
```

`is` (Kotlin)
```kotlin
if (mykey is VaultTransitAsymmetricKeyDetail) {
  for (version in mykey.versions) {
    ... 
  }
}
```

##### Example for explicit field method in Java  & Kotlin

```java
if (mykey.kind == VaultTransitKeyDetail.Kind.Asymmetric) {
  var myAsymKey = (VaultTransitAsymmetricKeyDetail) mykey;
  for (var version : myAsymKey.versions) {
    ...
  }
}
```

```kotlin
if (mykey.kind == VaultTransitKeyDetail.Kind.Asymmetric) {
  var myAsymKey = mykey as VaultTransitAsymmetricKeyDetail;
  for (version in myAsymKey.versions) {
    ...
  }
}
```

kdubb replied:
One final argument for subclasses, they're more future proof.  As Vault API grows and changes we can now distinctly add properties to either symmetric or asymmetric keys and their versions.

vsevel replied:
I agree with using a hierarchy. but I am not convinced that we should have it on the detail, plus the version. I am thinking that having it on the version is enough, which follows vault's api where the distinction is only made in the `keys` field.
could we get away with a single class `VaultTransitKeyDetail`, and the 3 classes for the version: `VaultTransitKeyVersion`, `VaultTransitAsymmetricKeyVersion` and `VaultTransitSymmetricKeyVersion`?
then the `instanceof` is only happening if you manipulate asymmetric keys, since for symmetric you can get the creation time from the base class anyway.
or do you think that it is important to make this distinction in the detail already?

kdubb replied:
I like having the detail be distinct for two reasons. First, you do an `instanceof` check and cast once on the detail and can now assume all the version objects are the same and can be used without casting; alternatively you'll cast each version in a loop. Second, it's future proofing for other areas where symmetric and asymmetric keys may differ in the future. Vault Transit is new and changing rapidly (as evidenced by these changes) so we could be required to break them out in the future regardless.

To your last point, remember if you only want the `creationTime`, `VaultTransitKeyDetail` users can get that in the current implementation without checking or downcasting because it's `versions` property is `Map<String, ? extends VaultTransitKeyVersion>`.

vsevel replied:
> First, you do an instanceof check and cast once on the detail and can now assume all the version objects are the same and can be used without casting

agreed. but that is the only value added by those subclasses, as they will be empty once you start using generics.

> alternatively you'll cast each version in a loop

only if you know it is asymmetric, and you need the extra information (beside the creation time), of if you do not know the type (and usually you do I suspect).

> Second, it's future proofing for other areas where symmetric and asymmetric keys may differ in the future

if they differ later, we will always be able to refactor them into the hierarchy you are proposing today.

to progress on this particular point, could you start parametrizing the detail base class with generics? this should empty out the subclasses, and help reason later about whether or not we want to drop the detail hirerachy altogether. thanks.

kdubb replied:
I gentrified `VaultTransitKeyDetail` to include the version class as a type parameter. 

A nice side-effect of this is that it announces to users that each version entry is always of the same exact subtype.

sberyozkin commented:
> Vault Transit docs still say it's in preview. Given that we can nix backward compatibility,

IMHO it is not sound

kdubb commented:
The docs for this the Vault Transit extension explicitly states...

> In preview, backward compatibility and presence in the ecosystem is not guaranteed. Specific improvements might require to change configuration or APIs and plans to become stable are under way. Feedback is welcome on our mailing list or as issues in our GitHub issue tracker.

Given that users have been (and continue to be) warned about possible breaking API changes. It seems now is the time to remove the excess properties; _before_ the API is stabilized with redundant data.

sberyozkin commented:
> Given that users have been (and continue to be) warned about possible breaking API changes. It seems now is the time to remove the excess properties; before the API is stabilized with redundant data.

But we are in a Quarkus land here, deprecating it and then removing it in for ex 2.1.0 would be better IMHO

kdubb commented:
@sberyozkin The docs I quoted are from Quarkus' documentation on using the Vault Transit Engine in Quarkus; not the Vault API docs.  You can see them located here https://quarkus.io/guides/vault-transit (the first info admonition details the "preview" status).

sberyozkin commented:
@kdubb 

> The docs I quoted are from Quarkus' documentation on using the Vault Transit Engine in Quarkus

LOL. My bad. If Vincent decides to drop it now then it is fine

kdubb commented:
The failing native tests are due to issues with explicit Jackson deserializers. I've created PR #17049 to fix the issue.

vsevel commented:
> The failing native tests are due to issues with explicit Jackson deserializers. I've created PR #17049 to fix the issue.

the other option is to ask jackson to map this to an `Object`. we should then get a simple integer for symmetric, and map for asymmetric key versions. then we can apply a bit of simple logic to build the appropriate object. I don't see the custom deserializer adding a lot of value here.

vsevel commented:
> @kdubb
> 
> > The docs I quoted are from Quarkus' documentation on using the Vault Transit Engine in Quarkus
> 
> LOL. My bad. If Vincent decides to drop it now then it is fine

@sberyozkin are we working against 1.x or 2.x here?
I have seen a comment at some point in quarkus dev that some backward incompatibilities could be baked into minor releases, to avoid a huge gap when going from 1.x to 2.x. [Migration Guides](https://github.com/quarkusio/quarkus/wiki/Migration-Guides) also show that at each version, we do have examples of such incompatibilities.

however in this particular case, since we reimplement a brand new field `versions`, we could just provide a new implementation of `getKeys` that rebuilds the `Map<String, Integer>` of the `versions` by navigating to `VaultTransitKeyVersion.creationTime` transformed back into `epoch`. then mark `getKeys` as deprecated. if we are working against 1.x then we can comment that this method will be removed in 2.x. if we are working against 2.x, just say that it will remove in a subsequent release.

this will avoid having to document the breaking change now, and is not hurting the code.

what do you all think? would that work for you @sberyozkin?

vsevel commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultTransitManager.java:
```diff
@@ -376,20 +383,55 @@ public VaultTransitKeyDetail readKey(String keyName) {
+                    .collect(Collectors.toMap(
+                            Entry::getKey,
+                            entry -> {
+                                VaultTransitKeyVersionData keyVersionData = entry.getValue();
```
please create dedicated methods such as `createAsymmetricKeyVersio(VaultTransitKeyVersionData)` so that we can have one liners lambdas.
as discussed elsewhere, I am not convinced by the `VaultTransitSymmetricKeyDetail` and `VaultTransitAsymmetricKeyDetail` classes.

kdubb commented:
@vsevel The `getKeys()` backward compatibility is already implemented and `getKeys()` is marked deprecated. I'm fine with leaving this as is for now. I guess we should probably choose a version where we switch from deprecation to removal.

kdubb commented:
@vsevel With regard to the custom deserializer, I'm ok with removing it but it _should_ work as Quarkus claims to support it (and after the PR it should be fine).

My only argument for leaving it in is that it makes life a lot easier as the Vault Transit API grows and changes. With the deserializer in place we only need to add new fields and things like nulls coupled with conversions (e.g. time) are handled more gracefully.

vsevel commented:
> @vsevel The `getKeys()` backward compatibility is already implemented and `getKeys()` is marked deprecated. I'm fine with leaving this as is for now. I guess we should probably choose a version where we switch from deprecation to removal.

yes I missed it. I was expecting the field `keys` and the setter to be dropped, then implementing the mapping logic directly into `getKeys`. but that's ok.
could you add a deprecated comment stating that the client code should use `getVersions()` instead, specifically on the `getKeys()` method?

kdubb commented:
I've documented `getKeys()` and it's depreciation as well as `getVersions()`.  I also followed your suggestion and removed the `keys` field altogether and implemented `getKeys()` as a computed property.

The only open questions are the "`VaultTransitKeyDetail` hierarchy" and "removing the deserializer".

vsevel commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/VaultTransitSecretEngine.java:
```diff
@@ -259,7 +259,7 @@
      * @see <a href="https://www.vaultproject.io/api-docs/secret/transit#read-key">read key</a>
      */
-    VaultTransitKeyDetail readKey(String keyName);
+    VaultTransitKeyDetail<?> readKey(String keyName);
```
? extends VaultTransitKeyVersion

vsevel replied:
you are correct. my bad. I did not think the compiler would accept:
```
        VaultTransitKeyDetail<?> vaultTransitKeyDetail = transit.readKey(KEY_NAME);
        VaultTransitKeyVersion version = vaultTransitKeyDetail.getVersions().get("1");
```
let's leave it as is.

vsevel commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultTransitManager.java:
```diff
@@ -365,7 +365,7 @@ public VaultTransitKeyExportDetail exportKey(String keyName, VaultTransitExportK
 
     @Override
-    public VaultTransitKeyDetail readKey(String keyName) {
+    public VaultTransitKeyDetail<?> readKey(String keyName) {
```
? extends VaultTransitKeyVersion

vsevel commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultTransitManager.java:
```diff
@@ -382,9 +382,9 @@ public VaultTransitKeyDetail readKey(String keyName) {
     }
 
-    protected VaultTransitKeyDetail map(VaultTransitReadKeyData data) {
+    protected VaultTransitKeyDetail<?> map(VaultTransitReadKeyData data) {
```
? extends VaultTransitKeyVersion

vsevel commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultTransitManager.java:
```diff
@@ -382,9 +382,9 @@ public VaultTransitKeyDetail readKey(String keyName) {
+    protected VaultTransitKeyDetail<?> map(VaultTransitReadKeyData data) {
         VaultTransitKeyVersionData latestVersionData = data.keys.get(Integer.toString(data.latestVersion));
-        VaultTransitKeyDetail result;
+        VaultTransitKeyDetail<?> result;
```
? extends VaultTransitKeyVersion

vsevel commented on integration-tests/vault/src/test/java/io/quarkus/vault/VaultTransitITCase.java:
```diff
@@ -324,7 +324,7 @@ public void asymmetricReadECDSAKey() {
         assertTrue(transitSecretEngine.listKeys().contains(KEY_NAME));
 
-        VaultTransitKeyDetail mykey = transitSecretEngine.readKey(KEY_NAME);
+        VaultTransitKeyDetail<?> mykey = transitSecretEngine.readKey(KEY_NAME);
```
? extends VaultTransitKeyVersion

vsevel commented on integration-tests/vault/src/test/java/io/quarkus/vault/VaultTransitITCase.java:
```diff
@@ -361,7 +361,7 @@ public void asymmetricReadRSAKey() {
         assertTrue(transitSecretEngine.listKeys().contains(KEY_NAME));
 
-        VaultTransitKeyDetail mykey = transitSecretEngine.readKey(KEY_NAME);
+        VaultTransitKeyDetail<?> mykey = transitSecretEngine.readKey(KEY_NAME);
```
? extends VaultTransitKeyVersion

vsevel commented on integration-tests/vault/src/test/java/io/quarkus/vault/VaultTransitITCase.java:
```diff
@@ -398,7 +398,7 @@ public void symmetricReadAESKey() {
         assertTrue(transitSecretEngine.listKeys().contains(KEY_NAME));
 
-        VaultTransitKeyDetail mykey = transitSecretEngine.readKey(KEY_NAME);
+        VaultTransitKeyDetail<?> mykey = transitSecretEngine.readKey(KEY_NAME);
```
? extends VaultTransitKeyVersion

kdubb commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/VaultTransitKeyDetail.java:
```diff
@@ -27,7 +28,7 @@ public String getName() {
     }
 
-    public VaultTransitKeyDetail setName(String name) {
+    public VaultTransitKeyDetail<KEYVER> setName(String name) {
```
To remove warnings about raw type usage.

vsevel replied:
I see.

kdubb commented:
@vsevel Checks are passing so I believe everything is good to go, save for the decision on detail hierarchy.

FYI, I've been using these changes to implement Vault based certificate signing and JWT signing with keys broadcasted as JWKs. Using it in a real scenario has been straightforward and pleasant.

vsevel commented:
can you actually move the transit dtos classes from package `io.quarkus.vault` to `io.quarkus.vault.transit`?
I realized that actually `VaultTransitExportKeyType` and `VaultTransitKeyDetail` were wrongly added to the base package. they should have been located in the `transit` sub-package.
so can you move those classes:
```
VaultTransitAsymmetricKeyDetail
VaultTransitAsymmetricKeyVersion
VaultTransitExportKeyType
VaultTransitKeyDetail
VaultTransitKeyVersion
VaultTransitSymmetricKeyDetail
VaultTransitSymmetricKeyVersion
```

basically `VaultTransit*` expect the interface `VaultTransitSecretEngine`.
this will be a breaking change for `VaultTransitExportKeyType` and `VaultTransitKeyDetail` since we are changing the package name, but I think this is a bug and we should fix it. ok with that @sberyozkin?

I saw that you left the deserializer, and the tests passed. I suppose you rebased on master that got your other PR merged? right?

@sberyozkin are you comfortable with the hierarchy? we are left with 2 empty classes `VaultTransitAsymmetricKeyDetail` and `VaultTransitSymmetricKeyDetail`, which essentially serve as typed markers for the version classes. this allows to do a cast on the detail instance, then be strongly typed on the versions as discussed previously. 

> FYI, I've been using these changes to implement Vault based certificate signing and JWT signing with keys broadcasted as JWKs. Using it in a real scenario has been straightforward and pleasant.

great feedback!

kdubb commented:
@vsevel  

> can you actually move the transit dtos classes from package io.quarkus.vault to io.quarkus.vault.transit?

Done. Types have been moved.

>I saw that you left the deserializer, and the tests passed. I suppose you rebased on master that got your other PR merged? right?

Yes. 😉

kdubb commented:
@vsevel I'd like to get this in before the next 2.x alpha/beta so we can use it without a custom branch.

 I know you have an outstanding question about the perceived added complexity of two hierarchies but I'd like to champion the idea of leaving it as is and merging it ASAP.

First, neither hierarchy is visible to most users and to those that will see it they will only see the detail hierarchy due to its new generic type parameter for key versions.  Second, it's forward looking and makes changes easier as we go forward. Finally, it's implemented and tested 😉

vsevel commented:
I am going to validate as is. if you are fine with the design @sberyozkin I will let you merge, else we will reopen the discussion.
@kdubb could you rebase to get the vault code from your other PR, which has been merged?
thanks for this contribution.

kdubb commented:
@vsevel @sberyozkin squashed and rebased to tip of main

sberyozkin commented:
@vsevel @kdubb Sorry I missed a few comments - looks like a nice compromise, thanks.
